File: advisor/suppressions.py
"""Targeted false-positive suppressions.

Baselines suppress everything-at-a-point-in-time; suppressions are
targeted per-rule, per-file, with optional expiry dates. They live in
``.advisor/suppressions.jsonl`` (JSON Lines) so the core package stays
zero-dep — no YAML runtime requirement. Each line is a JSON object with
the same fields a YAML entry would have.

## Schema (one JSON object per line, plus a top-line header)

    {"__advisor_suppressions__": true, "schema_version": "1.0"}
    {"rule_id": "advisor/high/abc1234567", "file": "src/legacy/parser.py",
     "reason": "rewrite scheduled for Q3", "until": "2026-09-01"}
    {"rule_id": "advisor/medium/def8901234", "file_glob": "tests/**/*.py",
     "reason": "test fixtures intentionally permissive"}

## Enforcement

- ``until`` is **required** for any suppression that applies to a
  finding above MEDIUM severity — expiry prevents "suppress and
  forget". Loading a file with a non-expiring > MEDIUM entry raises
  ``ValueError``.
- ``reason`` is required whenever ``until`` is required. Non-empty.
- An expired ``until`` logs a warning at load time and the entry is
  *not* silently re-activated — it must be renewed explicitly.
- ``file`` (exact path) and ``file_glob`` (fnmatch-style) are mutually
  exclusive. ``file_glob`` supports ``**``.

The loader is pure. Applying suppressions to a list of findings happens
via :func:`apply_suppressions`.
"""

from __future__ import annotations

import fnmatch
import json
import re
import warnings
from dataclasses import dataclass
from datetime import date, datetime, timezone
from pathlib import Path, PurePath

from .verify import Finding

SCHEMA_VERSION = "1.0"
_HEADER_KEY = "__advisor_suppressions__"

_SEVERITY_RANK: dict[str, int] = {
    "LOW": 1,
    "MEDIUM": 2,
    "HIGH": 3,
    "CRITICAL": 4,
}
_REQUIRES_EXPIRY_ABOVE = 2  # > MEDIUM requires an ``until`` date


@dataclass(frozen=True, slots=True)
class Suppression:
    """A single suppression entry."""

    rule_id: str
    reason: str
    file: str | None = None
    file_glob: str | None = None
    until: str | None = None  # ISO date, YYYY-MM-DD
    expired: bool = False  # set by the loader when ``until`` is in the past

    def matches(self, finding_path: str, finding_rule_id: str) -> bool:
        if finding_rule_id != self.rule_id:
            return False
        if self.file is not None and finding_path != self.file:
            return False
        if self.file_glob is not None and not _matches_glob(finding_path, self.file_glob):
            return False
        return True


def _matches_glob(file_path: str, pattern: str) -> bool:
    """Match ``file_path`` against ``pattern``, supporting ``**`` recursion."""
    p = PurePath(file_path).as_posix()
    if "**" in pattern:
        # Simple translation: ** → .*, * → [^/]*, ? → [^/]
        parts: list[str] = []
        i = 0
        while i < len(pattern):
            c = pattern[i]
            if c == "*":
                if i + 1 < len(pattern) and pattern[i + 1] == "*":
                    if i + 2 < len(pattern) and pattern[i + 2] == "/":
                        parts.append("(?:.*/)?")
                        i += 3
                    else:
                        parts.append(".*")
                        i += 2
                else:
                    parts.append("[^/]*")
                    i += 1
            elif c == "?":
                parts.append("[^/]")
                i += 1
            else:
                parts.append(re.escape(c))
                i += 1
        try:
            return bool(re.match("^" + "".join(parts) + "$", p))
        except re.error:
            # Malformed regex translation — fall through to plain fnmatch
            # so a suppressions.jsonl with an edge-case glob doesn't crash
            # the whole load pass.
            pass
    return fnmatch.fnmatch(p, pattern)


_DATE_ONLY_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def _parse_until(raw: str | None, *, context: str) -> tuple[str | None, bool]:
    """Return (normalized_until, expired) or raise ValueError on bad shape."""
    if raw is None or raw == "":
        return None, False
    # Reject datetime-shaped strings up front. Python 3.11+ ``date.fromisoformat``
    # accepts full ISO datetimes (e.g. ``2026-09-01T00:00:00+05:30``) and
    # silently drops the time and timezone, which would store an
    # off-by-one expiry under any non-UTC offset.
    if not _DATE_ONLY_RE.match(raw):
        raise ValueError(
            f"{context}: invalid until={raw!r}: expected YYYY-MM-DD (date only, no time/tz)"
        )
    try:
        d = date.fromisoformat(raw)
    except ValueError as exc:
        raise ValueError(f"{context}: invalid until={raw!r}: {exc}") from exc
    today = datetime.now(timezone.utc).date()
    expired = d < today
    return d.isoformat(), expired


def load_suppressions(path: Path) -> tuple[Suppression, ...]:
    """Load suppressions from a JSONL file.

    Missing file → empty tuple. Malformed lines raise :class:`ValueError`
    with a pointer to the offending line; entries that fail the
    "> MEDIUM requires ``until`` + ``reason``" rule raise the same. Use
    a try/except at the call site if you want graceful fallback.

    Expired ``until`` dates are not a hard error — they emit a warning
    and the matching entry keeps ``expired=True`` so callers can choose
    to respect-but-report rather than silently re-activate the finding.
    """
    if not path.exists():
        return ()
    try:
        text = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise ValueError(f"could not read {path}: {exc}") from exc

    entries: list[Suppression] = []
    for line_no, raw in enumerate(text.splitlines(), 1):
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        try:
            obj = json.loads(stripped)
        except json.JSONDecodeError as exc:
            raise ValueError(f"{path}:{line_no}: invalid JSON: {exc}") from exc
        if obj.get(_HEADER_KEY):
            version = str(obj.get("schema_version", ""))
            if version and version != SCHEMA_VERSION:
                warnings.warn(
                    f"{path}: schema_version {version!r} does not match "
                    f"expected {SCHEMA_VERSION!r}; parsing anyway",
                    UserWarning,
                    stacklevel=2,
                )
            continue
        ctx = f"{path}:{line_no}"
        rule_id = obj.get("rule_id")
        reason = obj.get("reason", "")
        file_ = obj.get("file")
        file_glob = obj.get("file_glob")
        until_raw = obj.get("until")

        if not rule_id or not isinstance(rule_id, str):
            raise ValueError(f"{ctx}: missing required 'rule_id'")
        if file_ and file_glob:
            raise ValueError(f"{ctx}: 'file' and 'file_glob' are mutually exclusive")
        if not (file_ or file_glob):
            raise ValueError(f"{ctx}: one of 'file' or 'file_glob' is required")

        until_norm, expired = _parse_until(until_raw, context=ctx)

        # Above-MEDIUM requires both until and reason.
        if _severity_from_rule_id(rule_id) > _REQUIRES_EXPIRY_ABOVE:
            if not until_norm:
                raise ValueError(
                    f"{ctx}: rule {rule_id!r} is above MEDIUM — 'until' date is required"
                )
            if not reason or not str(reason).strip():
                raise ValueError(
                    f"{ctx}: rule {rule_id!r} is above MEDIUM — non-empty 'reason' is required"
                )

        if expired:
            warnings.warn(
                f"{ctx}: suppression for {rule_id!r} expired on {until_norm}; renew or remove",
                UserWarning,
                stacklevel=2,
            )

        entries.append(
            Suppression(
                rule_id=str(rule_id),
                reason=str(reason or ""),
                file=str(file_) if file_ else None,
                file_glob=str(file_glob) if file_glob else None,
                until=until_norm,
                expired=expired,
            )
        )
    return tuple(entries)


def _severity_from_rule_id(rule_id: str) -> int:
    """Extract severity rank from a ``advisor/<sev>/<hash>``-shaped id.

    Foreign rule ids default to MEDIUM (rank 2) so expiry is not forced
    on callers that pipe findings from other tools.

    Forms recognized as authoritative-severity (subject to the
    ``until``/``reason`` gate above MEDIUM):

    * ``advisor/<sev>/<hash>`` — first segment is the namespace marker
    * ``<SEV>/<anything>`` — short form, e.g. ``HIGH/abc``; first
      segment IS the severity. Without this branch, ``HIGH/abc`` would
      decay to MEDIUM and bypass the expiry requirement.
    * Bare ``<SEV>`` — single segment treated as severity directly.
    """
    parts = rule_id.split("/")
    if parts and parts[0].upper() == "ADVISOR" and len(parts) >= 2:
        return _SEVERITY_RANK.get(parts[1].upper(), _REQUIRES_EXPIRY_ABOVE)
    if parts and parts[0].upper() in _SEVERITY_RANK:
        return _SEVERITY_RANK[parts[0].upper()]
    return _REQUIRES_EXPIRY_ABOVE


def apply_suppressions(
    findings: list[Finding],
    suppressions: tuple[Suppression, ...],
) -> tuple[list[Finding], list[tuple[Finding, Suppression]]]:
    """Drop findings that match an active (non-expired) suppression.

    Returns ``(kept, dropped_pairs)``. Expired suppressions are ignored
    for matching — their findings fall through to ``kept``.
    """
    from .sarif import synthesize_rule_id

    active = [s for s in suppressions if not s.expired]
    kept: list[Finding] = []
    dropped: list[tuple[Finding, Suppression]] = []
    for f in findings:
        rid = f.rule_id or synthesize_rule_id(f.severity, f.description)
        matched: Suppression | None = None
        for s in active:
            if s.matches(f.file_path, rid):
                matched = s
                break
        if matched is not None:
            dropped.append((f, matched))
        else:
            kept.append(f)
    return kept, dropped

File: src/suppressions.rs
//! Port of `advisor/suppressions.py` — targeted per-rule/per-file false-positive
//! suppressions with optional expiry. JSONL file under `.advisor/`.

use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::fs::{normalize_path, read_text_capped, MAX_ADVISOR_FILE_BYTES};
use crate::rank::{fnmatch_match, try_double_star_regex};
use crate::sarif::synthesize_rule_id;
use crate::Finding;

/// advisor's suppressions schema version.
pub const SCHEMA_VERSION: &str = "1.0";
const HEADER_KEY: &str = "__advisor_suppressions__";
const REQUIRES_EXPIRY_ABOVE: i64 = 2; // > MEDIUM requires `until`

fn severity_rank(sev: &str) -> Option<i64> {
    match sev {
        "LOW" => Some(1),
        "MEDIUM" => Some(2),
        "HIGH" => Some(3),
        "CRITICAL" => Some(4),
        _ => None,
    }
}

/// A single suppression entry. Mirrors `Suppression`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Suppression {
    pub rule_id: String,
    pub reason: String,
    pub file: Option<String>,
    pub file_glob: Option<String>,
    pub until: Option<String>,
    pub expired: bool,
}

impl Suppression {
    /// Whether this suppression matches a finding's path + rule id. Mirrors
    /// `Suppression.matches`.
    pub fn matches(&self, finding_path: &str, finding_rule_id: &str) -> bool {
        if finding_rule_id.to_uppercase() != self.rule_id.to_uppercase() {
            return false;
        }
        let np = normalize_path(finding_path);
        if let Some(file) = &self.file {
            if np != normalize_path(file) {
                return false;
            }
        }
        if let Some(glob) = &self.file_glob {
            if !matches_glob(&np, &normalize_glob_pattern(glob)) {
                return false;
            }
        }
        true
    }
}

/// Minimal-transform glob normalization (BOM strip + backslash→slash); does NOT
/// strip `:line` suffixes. Mirrors `_normalize_glob_pattern`.
fn normalize_glob_pattern(pattern: &str) -> String {
    pattern.trim_start_matches('\u{FEFF}').replace('\\', "/")
}

/// Match `file_path` against `pattern` with `**`-aware semantics; falls back to
/// fnmatch on a malformed pattern. Mirrors `_matches_glob`.
fn matches_glob(file_path: &str, pattern: &str) -> bool {
    match try_double_star_regex(pattern) {
        Some(re) => re.is_match(file_path),
        None => fnmatch_match(file_path, pattern),
    }
}

/// Extract severity rank from a rule id. Mirrors `_severity_from_rule_id`
/// (returns `Err` for a malformed advisor-namespace severity segment).
pub fn severity_from_rule_id(rule_id: &str) -> Result<i64, String> {
    let parts: Vec<&str> = rule_id.split('/').collect();
    if !parts.is_empty() && parts[0].to_uppercase() == "ADVISOR" && parts.len() >= 2 {
        let seg = parts[1].to_uppercase();
        return severity_rank(&seg).ok_or_else(|| {
            format!(
                "rule_id {rule_id:?} has malformed advisor-namespace severity segment {:?}; expected one of [\"CRITICAL\", \"HIGH\", \"LOW\", \"MEDIUM\"]",
                parts[1]
            )
        });
    }
    if !parts.is_empty() {
        if let Some(r) = severity_rank(&parts[0].to_uppercase()) {
            return Ok(r);
        }
    }
    Ok(REQUIRES_EXPIRY_ABOVE)
}

// ── date helpers (UTC) ─────────────────────────────────────────────

fn is_leap(y: i64) -> bool {
    (y % 4 == 0 && y % 100 != 0) || y % 400 == 0
}

fn days_in_month(y: i64, m: i64) -> i64 {
    match m {
        1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
        4 | 6 | 9 | 11 => 30,
        2 => {
            if is_leap(y) {
                29
            } else {
                28
            }
        }
        _ => 0,
    }
}

/// Days since 1970-01-01 for a proleptic-Gregorian date (Howard Hinnant's algorithm).
fn days_from_civil(y: i64, m: i64, d: i64) -> i64 {
    let y = if m <= 2 { y - 1 } else { y };
    let era = if y >= 0 { y } else { y - 399 } / 400;
    let yoe = y - era * 400;
    let doy = (153 * (if m > 2 { m - 3 } else { m + 9 }) + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146097 + doe - 719468
}

fn today_utc_days() -> i64 {
    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    (secs / 86400) as i64
}

/// Parse an `until` date (`YYYY-MM-DD`), returning `(normalized, expired)` or an
/// error on bad shape. Mirrors `_parse_until`.
fn parse_until(raw: Option<&str>, context: &str) -> Result<(Option<String>, bool), String> {
    let raw = match raw {
        None | Some("") => return Ok((None, false)),
        Some(s) => s,
    };
    // Date-only shape: ^\d{4}-\d{2}-\d{2}$
    let bytes = raw.as_bytes();
    let shape_ok = bytes.len() == 10
        && bytes[4] == b'-'
        && bytes[7] == b'-'
        && bytes[..4].iter().all(|b| b.is_ascii_digit())
        && bytes[5..7].iter().all(|b| b.is_ascii_digit())
        && bytes[8..10].iter().all(|b| b.is_ascii_digit());
    if !shape_ok {
        return Err(format!(
            "{context}: invalid until={raw:?}: expected YYYY-MM-DD (date only, no time/tz)"
        ));
    }
    let y: i64 = raw[..4].parse().unwrap_or(0);
    let m: i64 = raw[5..7].parse().unwrap_or(0);
    let d: i64 = raw[8..10].parse().unwrap_or(0);
    if !(1..=12).contains(&m) || d < 1 || d > days_in_month(y, m) {
        return Err(format!("{context}: invalid until={raw:?}: not a real date"));
    }
    let expired = days_from_civil(y, m, d) < today_utc_days();
    Ok((Some(raw.to_string()), expired))
}

/// Load suppressions from a JSONL file. Missing file → empty. Malformed JSON
/// lines are skipped; structural misuse returns `Err`. Mirrors `load_suppressions`.
pub fn load_suppressions(path: &Path) -> Result<Vec<Suppression>, String> {
    let text = match read_text_capped(path, MAX_ADVISOR_FILE_BYTES) {
        Ok(t) => t,
        Err(crate::fs::ReadCappedError::NotFound) => return Ok(Vec::new()),
        Err(e) => return Err(format!("could not read {}: {e}", path.display())),
    };

    let mut entries = Vec::new();
    for (i, raw_line) in text.lines().enumerate() {
        let line_no = i + 1;
        let stripped = raw_line.trim();
        if stripped.is_empty() || stripped.starts_with('#') {
            continue;
        }
        let obj: serde_json::Value = match serde_json::from_str(stripped) {
            Ok(v) => v,
            Err(err) => return Err(format!("{}:{}: invalid JSON — {err}", path.display(), line_no)),
        };
        let Some(map) = obj.as_object() else { continue };
        if map
            .get(HEADER_KEY)
            .and_then(|v| v.as_bool())
            .unwrap_or(false)
        {
            continue;
        }
        let ctx = format!("{}:{}", path.display(), line_no);
        let rule_id = map.get("rule_id").and_then(|v| v.as_str());
        let reason = map.get("reason").and_then(|v| v.as_str()).unwrap_or("");
        let file_ = map
            .get("file")
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty());
        let file_glob = map
            .get("file_glob")
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty());
        let until_raw = map.get("until").and_then(|v| v.as_str());

        let rule_id = match rule_id {
            Some(r) if !r.trim().is_empty() => r,
            _ => return Err(format!("{ctx}: missing required 'rule_id'")),
        };
        if file_.is_some() && file_glob.is_some() {
            return Err(format!(
                "{ctx}: 'file' and 'file_glob' are mutually exclusive"
            ));
        }
        if file_.is_none() && file_glob.is_none() {
            return Err(format!("{ctx}: one of 'file' or 'file_glob' is required"));
        }
        if let Some(glob) = file_glob {
            if try_double_star_regex(glob).is_none() {
                return Err(format!("{ctx}: invalid file_glob {glob:?}"));
            }
        }

        let (until_norm, expired) = parse_until(until_raw, &ctx)?;

        let sev_rank = severity_from_rule_id(rule_id).map_err(|e| format!("{ctx}: {e}"))?;
        if sev_rank > REQUIRES_EXPIRY_ABOVE {
            // Python `!r` uses single quotes around the rule id.
            if until_norm.is_none() {
                return Err(format!(
                    "{ctx}: rule '{rule_id}' is above MEDIUM — 'until' date is required"
                ));
            }
            if reason.trim().is_empty() {
                return Err(format!(
                    "{ctx}: rule '{rule_id}' is above MEDIUM — non-empty 'reason' is required"
                ));
            }
        }

        entries.push(Suppression {
            rule_id: rule_id.to_string(),
            reason: reason.to_string(),
            file: file_.map(|s| s.to_string()),
            file_glob: file_glob.map(|s| s.to_string()),
            until: until_norm,
            expired,
        });
    }
    Ok(entries)
}

/// Drop findings matching an active (non-expired) suppression. Returns
/// `(kept, dropped_pairs)`. Mirrors `apply_suppressions`.
pub fn apply_suppressions(
    findings: &[Finding],
    suppressions: &[Suppression],
) -> (Vec<Finding>, Vec<(Finding, Suppression)>) {
    let active: Vec<&Suppression> = suppressions.iter().filter(|s| !s.expired).collect();
    let mut kept = Vec::new();
    let mut dropped = Vec::new();
    for f in findings {
        let rid = match &f.rule_id {
            Some(r) if !r.is_empty() => r.clone(),
            _ => synthesize_rule_id(&f.severity, &f.description, "advisor"),
        };
        match active.iter().find(|s| s.matches(&f.file_path, &rid)) {
            Some(s) => dropped.push((f.clone(), (*s).clone())),
            None => kept.push(f.clone()),
        }
    }
    (kept, dropped)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    fn golden() -> Value {
        serde_json::from_str(include_str!("../tests/parity/suppressions.json")).unwrap()
    }

    #[test]
    fn severity_from_rule_id_matches_python() {
        let g = golden();
        for (rid, exp) in g["severity_from_rule_id"].as_object().unwrap() {
            assert_eq!(
                severity_from_rule_id(rid).unwrap(),
                exp.as_i64().unwrap(),
                "rid={rid}"
            );
        }
    }

    #[test]
    fn matches_match_python() {
        let g = golden();
        let m = &g["matches"];
        let s_exact = Suppression {
            rule_id: "advisor/low/a".into(),
            reason: "r".into(),
            file: Some("src/auth.py".into()),
            file_glob: None,
            until: None,
            expired: false,
        };
        let s_glob = Suppression {
            rule_id: "advisor/low/b".into(),
            reason: "r".into(),
            file: None,
            file_glob: Some("tests/**/*.py".into()),
            until: None,
            expired: false,
        };
        assert_eq!(
            s_exact.matches("src/auth.py:42", "advisor/low/a"),
            m["exact_hit"].as_bool().unwrap()
        );
        assert_eq!(
            s_exact.matches("src/other.py", "advisor/low/a"),
            m["exact_miss_path"].as_bool().unwrap()
        );
        assert_eq!(
            s_exact.matches("src/auth.py", "ADVISOR/LOW/A"),
            m["rule_case_insensitive"].as_bool().unwrap()
        );
        assert_eq!(
            s_glob.matches("tests/unit/test_x.py", "advisor/low/b"),
            m["glob_hit"].as_bool().unwrap()
        );
        assert_eq!(
            s_glob.matches("tests/test_x.py", "advisor/low/b"),
            m["glob_miss_shallow"].as_bool().unwrap()
        );
        assert_eq!(
            s_glob.matches("src/x.py", "advisor/low/b"),
            m["glob_miss_outside"].as_bool().unwrap()
        );
    }

    #[test]
    fn apply_matches_python() {
        let g = golden();
        let s_exact = Suppression {
            rule_id: "advisor/low/a".into(),
            reason: "r".into(),
            file: Some("src/auth.py".into()),
            file_glob: None,
            until: None,
            expired: false,
        };
        let findings = vec![
            Finding {
                file_path: "src/auth.py:1".into(),
                severity: "LOW".into(),
                description: "weak".into(),
                evidence: "e".into(),
                fix: "x".into(),
                rule_id: Some("advisor/low/a".into()),
                expected_vs_actual: String::new(),
            },
            Finding {
                file_path: "src/keep.py:1".into(),
                severity: "LOW".into(),
                description: "ok".into(),
                evidence: "e".into(),
                fix: "x".into(),
                rule_id: Some("advisor/low/z".into()),
                expected_vs_actual: String::new(),
            },
        ];
        let (kept, dropped) = apply_suppressions(&findings, &[s_exact]);
        let kp: Vec<String> = kept.iter().map(|f| f.file_path.clone()).collect();
        let dp: Vec<String> = dropped.iter().map(|(f, _)| f.file_path.clone()).collect();
        let exp = |k: &str| {
            g[k].as_array()
                .unwrap()
                .iter()
                .map(|v| v.as_str().unwrap().to_string())
                .collect::<Vec<_>>()
        };
        assert_eq!(kp, exp("apply_kept"));
        assert_eq!(dp, exp("apply_dropped"));
    }

    #[test]
    fn load_valid_matches_python() {
        let g = golden();
        let dir = std::env::temp_dir().join(format!("advisor_supp_{}", std::process::id()));
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("s.jsonl");
        std::fs::write(
            &path,
            "{\"__advisor_suppressions__\": true, \"schema_version\": \"1.0\"}\n\
             {\"rule_id\": \"advisor/low/a\", \"file\": \"src/auth.py\", \"reason\": \"ok\"}\n\
             {\"rule_id\": \"advisor/high/b\", \"file_glob\": \"legacy/**\", \"reason\": \"rewrite\", \"until\": \"2999-01-01\"}\n\
             {\"rule_id\": \"advisor/critical/c\", \"file\": \"x.py\", \"reason\": \"old\", \"until\": \"2000-01-01\"}\n",
        )
        .unwrap();
        let loaded = load_suppressions(&path).unwrap();
        let got: Vec<Value> = loaded
            .iter()
            .map(|s| {
                serde_json::json!({
                    "rule_id": s.rule_id, "file": s.file, "file_glob": s.file_glob,
                    "until": s.until, "expired": s.expired, "reason": s.reason,
                })
            })
            .collect();
        assert_eq!(Value::Array(got), g["load_valid"]);
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn load_errors_match_python() {
        let g = golden();
        let dir = std::env::temp_dir().join(format!("advisor_supp_err_{}", std::process::id()));
        std::fs::create_dir_all(&dir).unwrap();
        let run = |lines: &str| -> String {
            let path = dir.join("e.jsonl");
            std::fs::write(&path, lines).unwrap();
            let err = load_suppressions(&path).unwrap_err();
            // Strip the "path:line: " context prefix, like the golden generator.
            err.split_once(": ").map(|x| x.1.to_string()).unwrap_or(err)
        };
        assert_eq!(
            run("{\"file\": \"x.py\", \"reason\": \"r\"}\n"),
            g["err_missing_rule"].as_str().unwrap()
        );
        assert_eq!(run("{\"rule_id\":\"advisor/low/a\",\"file\":\"x.py\",\"file_glob\":\"y/**\",\"reason\":\"r\"}\n"), g["err_both"].as_str().unwrap());
        assert_eq!(
            run("{\"rule_id\":\"advisor/low/a\",\"reason\":\"r\"}\n"),
            g["err_no_target"].as_str().unwrap()
        );
        assert_eq!(
            run("{\"rule_id\":\"advisor/high/a\",\"file\":\"x.py\",\"reason\":\"r\"}\n"),
            g["err_high_no_until"].as_str().unwrap()
        );
        let _ = std::fs::remove_dir_all(&dir);
    }
}
